ios/adapters/epson/NetPrinterEpsonAdapter.m

```diff
@@ -1,112 +1,160 @@
+- (dispatch_queue_t)methodQueue {
+  return dispatch_get_global_queue(DISPATCH_QUEUE_PRIORITY_HIGH, 0);
 }
 
-- (void) connectAndSend:(NSString *)host
-               withPort:(nonnull NSNumber *)port
-           printRawData:(NSString *)text
-              epsonModel:(int)modelNumber
-                success:(RCTResponseSenderBlock)successCallback
-                   fail:(RCTResponseSenderBlock)errorCallback {
-    @try {
-        _retryAttempts=0;
-
-        printer = [[Epos2Printer alloc] initWithPrinterSeries:modelNumber lang:EPOS2_MODEL_ANK];
-        [printer setReceiveEventDelegate:self];
-        do{
-          @try{
-            _retryAttempts++;
-            [self connectAndSendAux:host printRawData:text success:successCallback fail:errorCallback];
-            _retryAttempts = 3;
-          }
-          @catch (NSException *exception)
-          {
-             if (_retryAttempts>=3)
-             {
-               @throw exception;
-             }
-             [printer endTransaction];
-             [printer disconnect];
-             [printer clearCommandBuffer];
-             [NSThread sleepForTimeInterval:(3.0f * _retryAttempts)];
-          }
-        }
-        while(_retryAttempts<3);
+- (void)connectAndSend:(NSString *)host
+              withPort:(nonnull NSNumber *)port
+          printRawData:(NSString *)text
+            epsonModel:(int)modelNumber
+               success:(RCTResponseSenderBlock)successCallback
+                  fail:(RCTResponseSenderBlock)errorCallback {
 
-    } @catch (NSException *exception) {
+  _retryAttempts = 0;
+  _finishedAsyncCall = NO;
 
-          [printer endTransaction];
-          int failResult=[printer disconnect];
-          [printer clearCommandBuffer];
-          [printer setReceiveEventDelegate:nil];
-          errorCallback(@[[NSString stringWithFormat:@"%@ and disconnect code %i",exception.reason,failResult]]);
-    }
-}
+  Epos2Printer *printer =
+      [[Epos2Printer alloc] initWithPrinterSeries:modelNumber
+                                             lang:EPOS2_MODEL_ANK];
+  NSString *printerLock;
 
-- (void) connectAndSendAux:(NSString *)host printRawData:(NSString *)text success:(RCTResponseSenderBlock)successCallback fail:(RCTResponseSenderBlock)errorCallback {
-        NSString* target = [NSString stringWithFormat:@"TCP:%@", host];
-        int result = EPOS2_SUCCESS;
-        result = [printer connect:target timeout:5000];
-        if (result != EPOS2_SUCCESS) {
-            [NSException raise:@"Invalid connection" format:@"Can't connect to printer %@, ERROR code: %i", host,result];
-        }
-
-        NSData* payload = [NSData dataWithBase64EncodedString:text];
+  @synchronized(printerDictionary) {
+    if (printerDictionary == nil) {
+      printerDictionary = [[NSMutableDictionary alloc] init];
+    }
+  }
 
-        [printer addCommand:payload];
-        result = [printer sendData:5000];
+  @synchronized(printerDictionary) {
```

Comment:
printerDictionary object is a global variable, hence its ref address is being used here as the ID for this code block. Any thread calling this code block of this ID, will lock, or wait until other the lock of `printerDictionary` is released.